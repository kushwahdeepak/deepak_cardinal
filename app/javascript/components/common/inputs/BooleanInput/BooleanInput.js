import React, { useState, useEffect } from 'react'
import isNil from 'lodash.isnil'
import isEmpty from 'lodash.isempty'

function BooleanInput({ label, setInputState, setFilter, testFunc, testAttr, checked, setCheckBox }) {

    useEffect(() => {
        if (!isNil(setFilter)) {
            console.assert(
                !(isNil(testFunc) && isNil(testAttr)),
                'BooleanInput requires either `testFunc` or `testAttr` to be provided'
            )
            testFunc = testFunc ?? getDefaultTestFunc(testAttr)
            setFilter(
                testAttr,
                !checked ? null : (candidate) => testFunc(candidate)
            )
        }

        if (!isNil(setInputState)) {
            setInputState(testAttr, checked)
        }
    }, [checked])

    const handleCheckboxChange = (e) => {
        setCheckBox(testAttr, e.target.checked)
    }

    return (
        <div className="d-flex align-items-center">
            <input
                checked={checked}
                id={testAttr}
                type="checkbox"
                data-test="checkbox-item"
                onChange={handleCheckboxChange}
                style={{ marginRight: '0.5rem' }}
            />
            <label
                data-test="label-item"
                style={{ marginBottom: '0' }}
                htmlFor={testAttr}
            >
                {label}
            </label>
        </div>
    )
}

function getDefaultTestFunc(testAttr) {
    return (candidate) => {
        return !isEmpty(candidate[testAttr])
    }
}

export default BooleanInput
