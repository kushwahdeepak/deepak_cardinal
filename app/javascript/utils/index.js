import { act } from 'react-dom/test-utils'

export const findByTestAttr = (wrapper, attr) => {
    return wrapper.find(`[data-test='${attr}']`)
}

export const timeoutResolve = (time = 0, response = {}) =>
    new Promise((resolve) => {
        setTimeout(() => resolve(response), time)
    })

export const wait = (time = 0) => {
    const timer = timeoutResolve(time)

    return act(async () => {
        await timer
    })
}
export function emailIsValid(email) {
    return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email)
}

export function generateRandom(min, max) {
    return Math.floor(min + Math.random() * (max + 1 - min))
}

export function capitalizeFirstLetter(string) {
    return string.charAt(0).toUpperCase() + string.slice(1).toLowerCase()
}

export function createMarkedList(array) {
    if (!Array.isArray(array)) {
        return []
    }
    return array.map((candidate) => ({
        id: candidate.id,
        name: candidate.id,
        fullName: `${candidate.first_name} ${candidate.last_name}`,
        emailAddress: candidate.email_address,
        isChecked: false,
    }))
}

export function stringReplace(string,toSymbol,fromSymbol){
    return string.replace(toSymbol, fromSymbol)
}

export function capitalize(string) {
    return string.replace(/\b(\w)/g, s => s.toUpperCase());
}

export function skillIsNotEmpty(value){
    return value.trim().length > 0;
}

export function firstCharacterCapital(value){
    return value?.charAt(0)?.toUpperCase() + value?.slice(1)
}
