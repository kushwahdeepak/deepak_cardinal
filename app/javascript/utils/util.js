import isNil from 'lodash.isnil'
import isEmpty from 'lodash.isempty'

class Util {
    /**
     * Define stage values
     */
    static STAGES = ['first_interview', 'second_interview', 'offer', 'reject']

    /**
     * Tests whether a single object attribute includes the at
     * least one string from the given set.
     */
    static attributePassesStringSetFilter(object, attr, stringSet) {
        if (isNil(object[attr])) return false

        // We "OR" the tag tests: the attribute needs to contain only one tag
        return stringSet
            .map((string) => string.trim().toLowerCase())
            .reduce(
                (passes, string) =>
                    passes ||
                    Util.attributeToString(object[attr])
                        .toLowerCase()
                        .includes(string),
                false
            )
    }

    /**
     * Tests whether at least one of a set of object attributes
     * includes every word from a given piece of text.
     */
    static objectPassesTextFilter(object, attrs, text) {
        text = text.trim().toLowerCase()

        if (text === '') return true

        const words = text.split(' ')

        // Join all the given attribute names to lowercase, stringified attribute values
        const allAttrsString = attrs
            .filter((attr) => !isNil(object[attr]))
            .map((attr) => Util.attributeToString(object[attr]).toLowerCase())
            .join('')

        // Is every word included in the allAttrsString?
        return words.reduce(
            (matchesAll, word) => matchesAll && allAttrsString.includes(word),
            true
        )
    }

    static attributeToString(attr) {
        if (typeof attr === 'string') {
            return attr
        } else if (Array.isArray(attr)) {
            return attr.join(',')
        } else if (typeof attr === 'object') {
            return JSON.stringify(attr)
        } else {
            return '' + attr
        }
    }

    static numberWithCommas(x) {
        return x?.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',')
    }

    /**
     * Returns number of results in format [x] - [y] of [z]
     * @param {number} totalJobsCount Total number of jobs
     * @param {number} pageCount Total number of pages
     * @param {number} activePage Currently active page
     * @param {number} pageSize How many results to display per page
     */
    static displayNumberOfResults(
        totalJobsCount,
        pageCount,
        activePage,
        pageSize,
        totalPersonCount=null,
        candidatePage=false,
        showdisplaycount=true
    ) {
        if (totalJobsCount == 0) return '0 - 0 of 0'
        const offset = pageSize * pageCount - totalJobsCount
        const lastJobIndex =
            (activePage + 1) * pageSize > totalJobsCount
                ? totalJobsCount
                : (activePage + 1) * pageSize
        const firstJobIndex =
            activePage == 0
                ? 1
                : activePage == pageCount - 1
                ? lastJobIndex - pageSize + offset + 1
                : lastJobIndex - pageSize + 1
        return `${firstJobIndex} - ${lastJobIndex} of ${Util.numberWithCommas(
            (candidatePage && showdisplaycount) ? totalPersonCount : totalJobsCount

        )}`
    }

    /**
     * Returns a string indicating the candidate's title and current
     * company, based on available information on the Person object
     * @param {object} candidate Person object from our db
     */
    static candidateCompanyString(candidate) {
        const position = candidate.company_position
        const currentCompany = candidate.current_company

        if (isEmpty(position) && isEmpty(currentCompany)) return ''

        if (!isEmpty(position) && !isEmpty(currentCompany))
            return `${position} at ${currentCompany}`

        if (!isEmpty(position)) return position

        if (!isEmpty(currentCompany)) return `at ${currentCompany}`
    }

    /**
     * Converts the given array of objects with ids to an object whose
     * keys are the ids of the array objects.
     *
     * Every object in the given array must have a unique 'id' property defined.
     *
     * @param {Array} array
     * @returns {Object}
     */
    static arrayToObject(array) {
        const obj = {}
        array.forEach((element) => (obj[element.id] = element))
        return obj
    }

    /**
     * Makes a copy of the given array and replace the value at 'index' with 'value'
     *
     * @param {Array} array
     * @param {Number} index
     * @param {*} value
     * @returns {Array}
     */
    static replaceInArrayCopy_i(array, index, value) {
        const clone = array.slice()
        clone[index] = value
        return clone
    }

    /**
     * Makes a copy of the given array and replaces the element with the same
     * .id property as obj.id with obj
     *
     * @param {Array} array
     * @param {Object} obj
     * @returns {Array}
     */
    static replaceInArrayCopy(array, obj) {
        const index = array.findIndex((element) => element.id === obj.id)
        return Util.replaceInArrayCopy_i(array, index, obj)
    }

    static deleteInArrayCopy(array, index) {
        const copy = array.slice()
        copy.splice(index, 1)

        return copy
    }

    static handleUndefined(value){
        if (value === undefined || value === 'undefined')
        return '';
        else
        return value
    }

    static handleUndefinedFullName(firstName, lastName){
      if(firstName === "undefined" || firstName === undefined)
      return lastName
      else if(lastName === "undefined" || lastName === undefined)
      return firstName
      else
      return firstName + ' ' + lastName     }

    static capitalize(value){
         return value.charAt(0).toUpperCase() + value.slice(1).toLowerCase();
    }
}

export default Util
