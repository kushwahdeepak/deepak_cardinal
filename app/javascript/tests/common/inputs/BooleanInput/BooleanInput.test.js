import React from 'react';
import BooleanInput from '../../../../components/common/inputs/BooleanInput/BooleanInput';
import renderer from 'react-test-renderer';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { findByTestAttr } from '../../../../utils';

configure({adapter: new Adapter()});

const initialShallowRender = (props = {}) => {
	return shallow(<BooleanInput {...props}/>);
};

const initialMountRender = (props = {}) => {
	return mount(<BooleanInput {...props}/>);
};

describe('BooleanInput component',  () => {
	let wrapper;
	let fullWrapper;

	beforeEach(() => {

		const props = {
		};
		wrapper = initialShallowRender(props);
		fullWrapper = initialMountRender(props);
	});

	afterEach(() => {
		fullWrapper.unmount();
	});

	it('should render BooleanInput component', () => {
		expect(wrapper.exists()).toBe(true);
	});

	it('should match with a snapshot', () => {
		const component = renderer.create(<BooleanInput />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	describe('BooleanInput component with props', () => {
		let wrapper;
		let fullWrapper;
		const setState = jest.fn();
		const useStateSpy = jest.spyOn(React, 'useState');
		useStateSpy.mockImplementation((init) => [init, setState]);

		beforeEach(() => {

			const props = {
				handleCheckboxChange: jest.fn(),
				testAttr: 'mockAttribute',
				setInputState: jest.fn(),
				label: 'mockLabel'
			};
			wrapper = initialShallowRender(props);
			fullWrapper = initialMountRender(props);
		});

		afterEach(() => {
			fullWrapper.unmount();
			jest.clearAllMocks();
		});

		it('should change checked for checkbox', () => {
			const checkboxItem = findByTestAttr(wrapper, 'checkbox-item').at(0);
			checkboxItem.simulate("change", { target: { checked: true } });
			expect(findByTestAttr(wrapper, 'checkbox-item').props().checked).toBe(true);
		});
		it('should call setInputState function', () => {
			expect(fullWrapper.props().setInputState).toHaveBeenCalledTimes(1);
		});

		it('should has correct testAttr prop', function () {
			const checkboxItem = findByTestAttr(wrapper, 'checkbox-item').at(0);
			expect(checkboxItem.props().id).toBe(fullWrapper.prop('testAttr'));
		});

		it('should has correct label text', function () {
			const labelItem = findByTestAttr(wrapper, 'label-item').at(0);
			expect(labelItem.text()).toBe(fullWrapper.prop('label'));
		});
	});
	describe('BooleanInput component without props', () => {

	});

});
