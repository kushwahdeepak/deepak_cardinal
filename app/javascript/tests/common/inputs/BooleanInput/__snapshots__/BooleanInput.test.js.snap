// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BooleanInput component should match with a snapshot 1`] = `
<div
  className="d-flex align-items-center"
>
  <input
    checked={false}
    data-test="checkbox-item"
    onChange={[Function]}
    style={
      Object {
        "marginRight": "0.5rem",
      }
    }
    type="checkbox"
  />
  <label
    data-test="label-item"
    style={
      Object {
        "marginBottom": "0",
      }
    }
  />
</div>
`;
